### PR TITLE
Password recovery

### PR DESCRIPTION
**Pull request by rubensayshi**

### CODE REVIEW

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+    angular.module("blocktrail.setup")
+        .controller("SetupChangePasswordCtrl", SetupChangePasswordCtrl);
+
+    function SetupChangePasswordCtrl($scope, $stateParams, $http, $q, $sce, PasswordStrength, dialogService, $log, $filter, sdkService, $translate, CONFIG) {
+        $scope.working  = false;
+        $scope.error    = null;
+        $scope.form     = {
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NIT: lets stop doing this aligning of `=` signs ;)  
not gonna fix it retroactively but we shouldn't continue this boaz-practice :P

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
...,"unit","universe","unknown","unlock","until","unusual","unveil","update","upgrade","uphold","upon","upper","upset","urban","urge","usage","use","used","useful","useless","usual","utility","vacant","vacuum","vague","valid","valley","valve","van","vanish","vapor","various","vast","vault","vehicle","velvet","vendor","venture","venue","verb","verify","version","very","vessel","veteran","viable","vibrant","vicious","victory","video","view","village","vintage","violin","virtual","virus","visa","visit","visual","vital","vivid","vocal","voice","void","volcano","volume","vote","voyage","wage","wagon","wait","walk","wall","walnut","want","warfare","warm","warrior","wash","wasp","waste","water","wave","way","wealth","weapon","wear","weasel","weather","web","wedding","weekend","weird","welcome","west","wet","whale","what","wheat","wheel","when","where","whip","whisper","wide","width","wife","wild","will","win","window","wine","wing","wink","winner","winter","wire","wisdom","wise","wish","witness","wolf","woman","wonder","wood","wool","word","work","world","worry","worth","wrap","wreck","wrestle","wrist","write","wrong","yard","year","yellow","you","young","youth","zebra","zero","zone","zoo"];
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: lets update the SDK today to expose it so it doesn't have to be here :D

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        var recoverySecret = null;
+        var walletVersion = null;
+        var requires2FA = null;
+
+        // Get state parameters
+        token = $stateParams.token;
+        recoverySecret = $stateParams.recovery;
+        walletVersion = $stateParams.version;
+        requires2FA = $stateParams.requires_2fa;
+        // ----
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: no need to seperate this from the `var token = null;, var recoverySecret = null;` etc

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+        var watchListener = $scope.$watch('form.ERS', function (newVal, oldVal) {
+            function strcmp(a, b) {
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: @romaroid can we put this somewhere in angular that makes sense to put small helper functions?

atm a bunch of these are at the bottom of `app.js`

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+                return (a < b ? -1 : ( a > b ? 1 : 0 ));
+            }
+
+            // Remove line breaks
+            if ($scope.form.ERS && $scope.form.ERS.length > 0) {
+                $scope.form.ERS = $scope.form.ERS.replace(/[\r\n]+/g, " "); // https://stackoverflow.com/a/34936253
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: @romaroid maybe `String.prototype.stripLineBreaks()` ? or that's not nice?

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                        if (err) {
+                            $log.error(err);
+                            $scope.alert({
+                                title: $translate.instant('ERROR'),
+                                body: "" + err
+                            });
+                        } else {
+                            pdf.save("BlockTrail Additional Recovery Data Sheet - " + identifier + ".pdf");
+
+                            // delete all temp backup info
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NIT: remove old comment

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+            return dialogService.prompt(
+                $translate.instant("SETUP_LOGIN"),
+                $translate.instant("MSG_MISSING_TWO_FACTOR_TOKEN")
+            ).result.then(function (result) {
+                twoFactorToken = result;
+            });
+        }
+
+        $scope.encryptNewERS = function() {
+
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NT: no blank line at start of function (I think also somewhere else)

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            var convert = function(s, from, to) {
+                return (new Buffer(s, from)).toString(to);
+            };
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: use `blocktrailSDK.convert` (and add `blocktrailSDK` as controller dependancy I think)

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+                var newEncryptedWalletSecretMnemonic;
+                var newPasswordHash;
+                var newPasswordBuffer = null;
+
+                if (walletVersion === blocktrailSDK.Wallet.WALLET_VERSION_V2) {
+                    newEncryptedSecret = blocktrailSDK.CryptoJS.AES.encrypt(secret, $scope.form.newPassword).toString(blocktrailSDK.CryptoJS.format.OpenSSL);
+                    newEncryptedWalletSecretMnemonic = blocktrailSDK.bip39.entropyToMnemonic(convert(newEncryptedSecret, 'base64', 'hex'));
+
+                } else {
+                    if (typeof $scope.form.newPassword === "string") {
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: when would `newPassword` not be a string?

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+                        newPasswordBuffer = new blocktrailSDK.Buffer($scope.form.newPassword);
+                    } else {
+                        if (!(newPasswordBuffer instanceof blocktrailSDK.Buffer)) {
+                            throw new Error('New password must be provided as a string or a Buffer');
+                        }
+                    }
+                    newEncryptedSecret = blocktrailSDK.Encryption.encrypt(secret, newPasswordBuffer);
+                    //newEncryptedSecret = blocktrailSDK.CryptoJS.AES.encrypt(accountInfo.secret, newPassword).toString()
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NIT: old line of comment

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+                    newEncryptedWalletSecretMnemonic = blocktrailSDK.EncryptionMnemonic.encode(newEncryptedSecret);
+                    // It's a buffer, so convert it back to base64
+                    newEncryptedSecret = newEncryptedSecret.toString('base64');
+                }//else
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NIT: we don't do this //endofablock stuff ;)

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+                    }
+
+                    // Try decrypting
+                    try {
+                        if (walletVersion === 'v3') {
+                            encryptedRecoverySecret = blocktrailSDK.EncryptionMnemonic.decode(encryptedRecoverySecretMnemonic);
+                            secret = blocktrailSDK.Encryption.decrypt(encryptedRecoverySecret, new blocktrailSDK.Buffer(recoverySecret, 'hex'));
+
+                        } else {
+                            // TODO: BIP39 not found :/
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: todo?

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+                } catch (e) {
+                    $scope.working = false;
+                    $log.error(e, e.message);
+                }
+            }
+        };
+    }
+
+    angular.module("blocktrail.setup").filter('filterERS', function () {
+
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: NIT: line break

File: src/js/modules/setup/controllers/change-password/change-password.tpl.html [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+                        <div class="form-group form-group-lg form-group-buttons">
+                            <button class="form-control btn btn-alt btn-primary" ng-disabled="working" ng-click="encryptNewERS()">{{ 'CONTINUE' | translate }}</button>
+                            <div ng-if="working">
+                                <loading-spinner></loading-spinner>
+                                {{ 'PLEASE_WAIT' | translate }}
+                            </div>
+                        </div>
+                        <!--<div class="smallButtons">-->
+                            <!--<a class="sentence-case" ui-sref="app.setup.login">{{ 'BACK' | translate }}</a>-->
+                        <!--</div>-->
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.tpl.html]: NIT: dont leave old code sitting around in comments, that's boaz-practice ;)

File: src/js/modules/setup/controllers/forgot-password/forgot-password.tpl.html [hunk 15]
```diff
@@ -1,19 +1,25 @@
-<div class="container">
+<div class="container" xmlns="http://www.w3.org/1999/html">
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/forgot-password/forgot-password.tpl.html]: dont think we need this `xmlns=` here ?

File: src/js/modules/setup/controllers/wallet-init/wallet-init.ctrl.js [hunk 16]
```diff
@@ -120,7 +120,8 @@
                             walletVersion: CONFIG.DEFAULT_WALLET_VERSION,
                             identifier: $scope.setupInfo.identifier,
                             password: $scope.setupInfo.password,
-                            support_secret: supportSecret
+                            support_secret: supportSecret,
+                            keyIndex: CONFIG.DEVKEYINDEX ? CONFIG.DEVKEYINDEX : 0
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/wallet-init/wallet-init.ctrl.js]: can be shortened to `CONFIG.DEVKEYINDEX || 0`

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+                        $log.error(e, e.message);
+                        $scope.working = false;
+                        return dialogService.alert(
+                            $translate.instant("RECOVERY_ERROR"),
+                            $translate.instant("MSG_PASSWORD_NOT_CHANGED")
+                        ).result;
+                    }
+
+                    watchListener();
+                    console.log(secret);
```
Comment (rubensayshi) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: strip `console.log` or release with `$log.debug`

also, why do we unsub the `watchListener` ? and maybe better to call that `watchListener` something like `unsubWatchERS`? @romaroid

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js (hunk 3 above)
Comment (lacksfish) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: So just assigning values right away can be done as well, if I understand correctly.

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js (hunk 8 above)
Comment (lacksfish) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: Done.

File: src/js/modules/setup/controllers/change-password/change-password.tpl.html (hunk 14 above)
Comment (lacksfish) [inline src/js/modules/setup/controllers/change-password/change-password.tpl.html]: Sorry :p

File: src/js/modules/setup/controllers/wallet-init/wallet-init.ctrl.js (hunk 16 above)
Comment (lacksfish) [inline src/js/modules/setup/controllers/wallet-init/wallet-init.ctrl.js]: Gotcha.

File: src/js/modules/setup/controllers/forgot-password/forgot-password.tpl.html (hunk 15 above)
Comment (lacksfish) [inline src/js/modules/setup/controllers/forgot-password/forgot-password.tpl.html]: Nope.

File: src/js/modules/setup/controllers/change-password/change-password.ctrl.js [hunk 18]
```diff
@@ -200,76 +215,77 @@
...
-                    }
-
-                    if (recoverySecret.length != 64) {
-                        $scope.working = false;
-                        return dialogService.alert(
-                            $translate.instant("RECOVERY_ERROR"),
-                            $translate.instant("MSG_CORRUPTED_SECRET")
-                        ).result;
-                    }
+                recoverySecret.then(function(recoverySecret) {
```
Comment (afk11) [inline src/js/modules/setup/controllers/change-password/change-password.ctrl.js]: minor nit- the promise and the resolved value are named the same

**Pull request merged**